examples/test_application/src/main.rs

```diff
@@ -0,0 +1,451 @@
+//! Example usage of [`tss_ecdsa`] crate.
+//!
+//! Each [`Participant`] is represented by a worker thread. This example uses
+//! standard library shared memory MPSC channels to communicate [`Message`]s
+//! amongst threads.
+
+use anyhow::{self};
+use clap::Parser;
+use rand::{rngs::OsRng, thread_rng};
+use sha2::{Digest, Sha256};
+use std::{
+    collections::HashMap,
+    sync::mpsc::{channel, Receiver, Sender},
+    thread,
+    time::{Duration, Instant},
+};
+use tracing::{debug, info, instrument, span, trace, Level};
+use tracing_subscriber::{self, EnvFilter};
+use tss_ecdsa::{
+    AuxInfoParticipant, AuxInfoPrivate, AuxInfoPublic, Identifier, KeySharePrivate, KeySharePublic,
+    KeygenParticipant, Message, Participant, ParticipantConfig, ParticipantIdentifier,
+    PresignInput, PresignParticipant, PresignRecord, ProtocolParticipant,
+};
+
+/// Messages from main thread to Participant worker threads.
+#[derive(Debug, Copy, Clone, Eq, Hash, PartialEq)]
+enum SubProtocol {
+    KeyGeneration,
+    AuxInfo,
+    Presign,
+    Sign,
+}
+
+#[derive(Parser, Debug)]
+#[command(author, version, about, long_about = None)]
+struct CommandLineArgs {
+    /// Number of participant worker threads to use.
+    #[arg(short, long)]
+    number_of_workers: usize,
+    /// Number of times to perform each sub-protocol of tss-ecdsa.
+    /// protocol_executions > 1 useful for computing meaningful average
+    /// execution times.
+    #[arg(short, long)]
+    protocol_executions: usize,
+}
+
+/// Outgoing message from worker threads to main thread.
+#[derive(Debug)]
+enum MessageFromWorker {
+    /// Worker made progress on sub-protocol, returning some messages.
+    FinishedRound(Vec<Message>),
+    /// Current sub-protocol ended.
+    SubProtocolEnded,
+}
+
+/// Maps [`ParticipantIdentifier`] to [`Sender`] channels for routing message to
+/// the correct participant.
+type WorkerChannels = HashMap<ParticipantIdentifier, Sender<Message>>;
+/// Channels for informing workers of the next [`SubProtocol`] to perform with
+/// the given SID [`Identifier`].
+type NextActionChannels = Vec<Sender<(SubProtocol, Identifier)>>;
+
+/// 1) Set up logging.
+/// 2) Create MPSC channels for communication between the workers and main
+/// thread. 3) Spawns N participant/worker threads ensuring the channels are
+/// "connected" properly. 4) Main thread executes the entire [`tss_ecdsa`]
+/// protocol to sign a message.
+fn main() -> anyhow::Result<()> {
+    let cli = CommandLineArgs::parse();
+    // Set up logging.
+    let filter = EnvFilter::from_default_env().add_directive("test_application=info".parse()?);
+    tracing_subscriber::fmt()
+        .with_env_filter(filter)
+        .without_time()
+        .compact()
+        .init();
+    let span = span!(Level::INFO, "main");
+    let _enter = span.entered();
+
+    let num_workers = cli.number_of_workers;
+    let rng = &mut OsRng;
```

Comment:
suggestion: `StdRng` is somewhat faster than going straight to the OS for randomness and would be suitable for this setting. I think the `StdRng::from_entropy` constructor gets its seed from OsRng anyway.